### ISSUE TITLE
Scribbler: convert blocks and system to C

### ISSUE THREAD

**Issue by MatzElectronics** [parallaxinc/BlocklyProp#824]
Necessary for:

Educational reasons
Simplify the blockly system overall
Makes blockly easier to maintain and improve

Possible becuase:
Eric Smith and The Amazing Andy!

**Comment by MatzElectronics** [comment]
Broken - hopefully fixed by #830

**Comment by MatzElectronics** [comment]
This is all ready for testing.  The only block that is known broken for now is the restore demo (factory default) block.  To access the s3 c system, choose new Propeller C project, then select Scribbler Robot as the board type.  The system looks identical, but is generating C instead.

**Comment by wstodghill** [comment]
Is the C "Scribbler Object" equivalent available somewhere?

**Comment by MatzElectronics** [comment]
@wstodghill Yes, but at this time it is not documented at all - you'd be looking at pure code.  You can see it here: https://github.com/parallaxinc/Simple-Libraries/tree/master/Learn/Simple%20Libraries/Robotics/S3/libs3

**Comment by wstodghill** [comment]
Thanks Matt!

**Comment by MatzElectronics** [comment]
Obstacle detection has failed so far - it's not reading at all.  Submitting an issue to Simple Libraries where the s3c object is hosted.

**Comment by MatzElectronics** [comment]
Simple line detection failed.
Send text to terminal failed.

**Comment by MatzElectronics** [comment]
All "If *sensor*" blocks are failing.  The issue is in the s3 c library, and I've narrowed it down.  Issue posted in Simple Libraries, and will implement a fix there soon.  No changes will likely be required in the blocklyprop system.

**Comment by MatzElectronics** [comment]
@zfi I figured out how to turn off the menu for "New Project" and just make it a link that takes you straight to making a new PropC project.  I'll put that into a PR when we are ready to dump the SPIN stuff completely.  I'd like to figure out how to set the icon based on board instead of language, and create a cool icon for each of the different boards (S3, PAB, HEB, and FLIP).

**Comment by zfi** [comment]
@MatzElectronics It looks like we need to change the drop-down object to a button object and then attach the "/blockly/projectcreate.jsp?lang=PROPC" URI to the ref tag to the new button anchor tag. 

There is a chuck of JavaScript in the community project page that enumerates the board names based upon the board type. It should be a fairly easy exercise to add a js function to map the board type to an image and then include that reference. [had facemail meeting]

**Comment by MatzElectronics** [comment]
@zfi @PropGit There was a library fix in SimpleIDE to address the latest bug here that may need to be pulled in if it's not automatic - Andy pulled a provisional repair into SimpleLibraries master, but it looks like it still needs to be merged into the BlocklyProp branch there.

**Comment by zfi** [comment]
I don't know about a SimpleIDE library. Maybe Simple-Libraries? The Demo server is fed from the BlocklyProp branch of that library. Andy should push a pr from the master branch to the BlocklyProp branch. Once that is done, the Demo stack will update automatically on the next quarter hour.

**Comment by MatzElectronics** [comment]
I thought that's how it worked...hard to keepntrack of it all :)

**Comment by MatzElectronics** [comment]
From @mrodriguez-parallax:

The code within the "if" blocks below executes whether or not the "if" condition is met or not.

![image](https://cloud.githubusercontent.com/assets/19737272/21405159/8778cffc-c779-11e6-8105-779da038c598.png)

This loop block in Simple Control does not loop at all. In the C code it looks like n is never less than 0 and thus doesn't start the loop.

![image](https://cloud.githubusercontent.com/assets/19737272/21405181/97ea9302-c779-11e6-9f2d-31b15a07fdad.png)

![image](https://cloud.githubusercontent.com/assets/19737272/21405200/aad1a852-c779-11e6-84ab-2cdd11e3bd83.png)

Last but not least I couldn't get the ping block to output any distance data.
![image](https://cloud.githubusercontent.com/assets/19737272/21405211/b7df031e-c779-11e6-9264-3f95a7a7b6a2.png)
Here's the code generated for the ping block above. I don't know what the s3.h  library does but the rest of the C code looks okay.
![image](https://cloud.githubusercontent.com/assets/19737272/21405221/c29690ce-c779-11e6-874c-85186838c89d.png)

I will keep on testing other blocks and let you know my findings. Let me know if you want me to put this in Github. I can't remember if Github is for the regular blockly or also the demo blockly.

**Comment by MatzElectronics** [comment]
From @mrodriguez-parallax's testing, I'll repair the loop and ping blocks as I can tell what's happening there.  The sensor conditionals might still be a Simple Libraries issue, so that may take a bit more work to figure out what is going wrong there.

**Comment by MatzElectronics** [comment]
@mrodriguez-parallax the PING and loop above should now be fixed.  Go ahead and test them and let me know.  I haven't figured out the sensors yet.  I'll let you know when I do.

**Comment by MatzElectronics** [comment]
@AndyLindsay @zfi Some progress here, but it's again a s3 C library issue.  The simple sensors blocks that @mrodriguez-parallax described above as always evaluating to true regardless of the sensors readings are caused by some conditional logic that was convoluted when the library was translated from SPIN.  

Now that I understand what is going on I can re-write the conditional sets and then submit them to Simple Libraries.  I've gotten the obstacle stuff completely rewritten and it's testing well in SimpleIDE, so I'll apply the same fix to the others.

**Comment by MatzElectronics** [comment]
Okay @mrodriguez-parallax @wstodghill!  This should now be ready to fully test (except for the "restore factory defaults" block!  Let me know if you find any issues - I'd love to close this one out!
Thanks!

**Comment by wstodghill** [comment]
@MatzElectronics - been playing a bit and no problems - but not a full-scale test yet! Looking good so for...

**Comment by MatzElectronics** [comment]
We are closer to this.  Testing status is in this google doc:
https://docs.google.com/spreadsheets/d/1uaZrnuFXr3wLSi_7DHNb4rNptnQnzexZBm40L79fphI/edit#gid=0

**Comment by zfi** [comment]
Updated the Cloud Compiler server to send a precompiled binary file to the client when an S3 restore_s3_defaults block is encountered. This is now available on the Demo server.

@mrodriguez-parallax Can you test this one please? Create an S3 project with a single 'restore_s3_defaults' block and verify that it does, in fact, load up the S3 demo program.

**Comment by wstodghill** [comment]
@zfi - Works in Louisiana!

**Comment by MatzElectronics** [comment]
Great, thank you!

*Matthew Matz  |  STEM/Robotics Educator*
Parallax Inc.  |  Direct: 916-625-3019  |  www.parallax.com  |  @M
atzElectronics <http://twitter.com/MatzElectronics>

**Comment by MatzElectronics** [comment]
Let's beat on the system for a while before we call this issue done...

**Comment by paragitadmin** [comment]
We were especially concerned with how the S3 reset was going to work when in close proximity to spicy foods. I see that it was no sweat. [I couldn't resist]

@MatzElectronics I see that if the restore_s3 block is embedded in a project with other blocks, the Cloud Compiler will disregard the program when it spots the restore block. In this situation, the client editor will receive the binary for reset and not the compiled program that the user might be expecting.

It is possible to add a Kryptonite block detector and advise the user when the editor encounters this set of conditions?

**Comment by MatzElectronics** [comment]
Yup!  It can throw a warn icon if it detects any other blocks.  That same
code will be useful elsewhere, too

**Comment by wstodghill** [comment]
Ha!

**Comment by MatzElectronics** [comment]
Closing.  This is all done, tested, working.  New requests can go in as new issues.
